core/src/main/scala/org/apache/spark/api/r/RRunner.scala

```diff
@@ -149,12 +150,23 @@ private[spark] class RRunner[U](
 
           dataOut.writeInt(numPartitions)
 
-          dataOut.writeInt(if (isDataFrame) 1 else 0)
+          val mode = if (isDataFrame && key != null) {
+            2 // gapply
+          } else if (isDataFrame) {
+            1 // dapply
+          } else {
+            0 // RDD
+          }
+          dataOut.writeInt(mode)
 
           if (isDataFrame) {
             SerDe.writeObject(dataOut, colNames)
           }
 
+          if (key != null) {
```

Comment:
It is the grouping columns(key) used for gapply mode. I can also additionally check for gapply mode before writing it.